contracts/masset/Masset.sol

```diff
@@ -304,44 +345,61 @@ contract Masset is
         (bool isValid, string memory reason, BassetDetails memory inputDetails, BassetDetails memory outputDetails) =
-            basketManager.prepareSwapBassets(_input, _output, false);
+            basketManager.prepareSwapBassets(args.input, args.output, false);
         require(isValid, reason);
 
+        Cache memory cache = _getCacheDetails();
+
         // 3. Deposit the input tokens
-        uint256 quantitySwappedIn = _depositTokens(_input, inputDetails.integrator, inputDetails.bAsset.isTransferFeeCharged, _quantity);
+        (uint256 amnountIn, ) =
+            _depositTokens(args.input, inputDetails.bAsset.ratio, inputDetails.integrator, inputDetails.bAsset.isTransferFeeCharged, _quantity, cache.maxCache);
         // 3.1. Update the input balance
-        basketManager.increaseVaultBalance(inputDetails.index, inputDetails.integrator, quantitySwappedIn);
+        basketManager.increaseVaultBalance(inputDetails.index, inputDetails.integrator, amnountIn);
 
         // 4. Validate the swap
         (bool swapValid, string memory swapValidityReason, uint256 swapOutput, bool applySwapFee) =
-            forgeValidator.validateSwap(totalSupply(), inputDetails.bAsset, outputDetails.bAsset, quantitySwappedIn);
+            forgeValidator.validateSwap(cache.vaultBalanceSum, inputDetails.bAsset, outputDetails.bAsset, amnountIn);
         require(swapValid, swapValidityReason);
+        require(swapOutput > 0, "Must withdraw something");
 
         // 5. Settle the swap
+        // 5.0. Redeclare recipient to avoid stack depth error
+        address recipient = args.recipient;
         // 5.1. Decrease output bal
-        basketManager.decreaseVaultBalance(outputDetails.index, outputDetails.integrator, swapOutput);
-        // 5.2. Calc fee, if any
-        if(applySwapFee){
-            swapOutput = _deductSwapFee(_output, swapOutput, swapFee);
-        }
-        // 5.3. Withdraw to recipient
-        IPlatformIntegration(outputDetails.integrator).withdraw(_recipient, _output, swapOutput, outputDetails.bAsset.isTransferFeeCharged);
-
-        output = swapOutput;
-
-        emit Swapped(msg.sender, _input, _output, swapOutput, _recipient);
+        Amount memory amt = _withdrawTokens(
+            WithdrawArgs({
+                quantity: swapOutput,
+                bAsset: args.output,
+                integrator: outputDetails.integrator,
+                feeRate: applySwapFee ? swapFee : 0,
+                hasTxFee: outputDetails.bAsset.isTransferFeeCharged,
+                recipient: recipient,
+                ratio: outputDetails.bAsset.ratio,
+                maxCache: cache.maxCache,
+                vaultBalance: outputDetails.bAsset.vaultBalance
+            })
+        );
+
+        basketManager.decreaseVaultBalance(outputDetails.index, outputDetails.integrator, amt.net);
+
+        surplus = cache.surplus.add(amt.scaledFee);
+
+        emit Swapped(msg.sender, args.input, args.output, amt.net, recipient);
```

Comment:
It seems like `output` wasn't assigned here.

Reply:
Sorry I don't understand

Reply:
It seems like the return value `output` is never assigned along this execution path. It would just be 0 always.

Reply:
Gotcha, resolved.